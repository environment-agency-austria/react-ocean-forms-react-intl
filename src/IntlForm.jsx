/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'react-ocean-forms';
import { injectIntl, intlShape } from 'react-intl';

/**
 * React-intl wrapper for the OceanJS
 * Forms component.
 */
class IntlForm extends React.Component {
  constructor(props) {
    super(props);

    this.formatIntlString = this.formatIntlString.bind(this);
  }

  formatIntlString(id, values) {
    // Return null if no valid id was given
    if (id === null || id === undefined || id === '') {
      console.error('[IntlForm] Invalid id given to formatIntlString');
      return null;
    }

    const { intl } = this.props;
    return intl.formatMessage({ id }, values);
  }

  render() {
    const {
      children,
      ...rest
    } = this.props;

    return (
      <Form
        {...rest}
        formatString={this.formatIntlString}
      >
        {children}
      </Form>
    );
  }
}

IntlForm.displayName = 'IntlForm';

IntlForm.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  intl: intlShape.isRequired,
};

export const BaseIntlForm = IntlForm;
export default injectIntl(IntlForm);
