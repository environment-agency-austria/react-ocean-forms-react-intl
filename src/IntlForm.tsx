/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { InjectedIntlProps, injectIntl } from 'react-intl';
import { Form, IFormProps, IMessageValues } from 'react-ocean-forms';

/**
 * Props for the IntlForm component
 */
export interface IIntlFormProps extends IFormProps, InjectedIntlProps { }

/**
 * React-intl wrapper for the OceanJS
 * Forms component.
 */
export class BaseIntlForm extends React.Component<IIntlFormProps> {
  public static displayName: string = 'IntlForm';

  private formatIntlString = (id: string, values: IMessageValues): string => {
    // Return the id if no valid id was given
    if (id === null || id === undefined || id === '') {
      // tslint:disable-next-line:no-console
      console.error('[IntlForm] Invalid id given to formatIntlString');

      return id;
    }

    const { intl } = this.props;

    return intl.formatMessage({ id }, values);
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      children,
      ...rest
    } = this.props;

    return (
      <Form
        {...rest}
        formatString={this.formatIntlString}
      >
        {children}
      </Form>
    );
  }
}

export const IntlForm = injectIntl(BaseIntlForm);
