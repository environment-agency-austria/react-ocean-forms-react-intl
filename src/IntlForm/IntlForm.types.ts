import { InjectedIntlProps } from 'react-intl';
import { Form, PropsOf } from 'react-ocean-forms';

type TFormProps = JSX.LibraryManagedAttributes<typeof Form, PropsOf<typeof Form>>;

/**
 * Props for the IntlForm component
 */
export interface IIntlFormProps extends TFormProps, InjectedIntlProps { }
