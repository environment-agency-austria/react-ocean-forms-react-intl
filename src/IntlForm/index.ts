export * from './IntlForm';
export * from './IntlForm.types';
