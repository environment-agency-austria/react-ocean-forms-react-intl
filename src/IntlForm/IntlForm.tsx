/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { injectIntl } from 'react-intl';
import { Form, IMessageValues } from 'react-ocean-forms';

import { IIntlFormProps } from './IntlForm.types';

/**
 * React-intl wrapper for the OceanJS
 * Forms component.
 */
export class BaseIntlForm extends React.Component<IIntlFormProps> {
  public static displayName: string = 'IntlForm';

  private formatIntlString = (id: string, values: IMessageValues): string => {
    // Return the id if no valid id was given
    if (id === null || id === undefined || id === '') {
      // tslint:disable-next-line:no-console
      console.error('[IntlForm] Invalid id given to formatIntlString');

      return id;
    }

    const { intl } = this.props;

    return intl.formatMessage({ id }, values);
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      children,
      ...rest
    } = this.props;

    return (
      <Form
        {...rest}
        formatString={this.formatIntlString}
      >
        {children}
      </Form>
    );
  }
}

export const IntlForm = injectIntl(BaseIntlForm);
