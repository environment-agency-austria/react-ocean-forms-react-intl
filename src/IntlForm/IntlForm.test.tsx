import * as React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { InjectedIntl } from 'react-intl';
import { TSTringFormatter } from 'react-ocean-forms';

import { BaseIntlForm } from './IntlForm';

describe('<IntlForm />', () => {
  interface ISetupResult {
    wrapper: ShallowWrapper;
    intl: InjectedIntl;
    formatIntlString: TSTringFormatter;
  }

  const setup = (): ISetupResult => {
    const intl: InjectedIntl = {
      formatDate: jest.fn(),
      formatTime: jest.fn(),
      formatRelative: jest.fn(),
      formatNumber: jest.fn(),
      formatPlural: jest.fn(),
      formatMessage: jest.fn().mockImplementation((value: { id: string }) => value.id),
      formatHTMLMessage: jest.fn(),
      locale: 'en',
      formats: null,
      messages: {},
      defaultLocale: 'en',
      defaultFormats: null,
      now: jest.fn(),
      onError: jest.fn(),
    };

    const wrapper = shallow((
      <BaseIntlForm
        intl={intl}
      />
    ));

    const formatIntlString = wrapper.find('Form').prop<TSTringFormatter>('formatString');

    return {
      wrapper,
      intl,
      formatIntlString,
    };
  };

  it('should render without crashing', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass a formatString method to the base form', () => {
    const { wrapper } = setup();
    expect(wrapper.find('Form').prop('formatString')).toBeInstanceOf(Function);
  });

  describe('formatIntlString', () => {
    it('should pass the message id to intl.formatMessage and return the result', () => {
      const { intl, formatIntlString } = setup();

      const mockMessage = 'mock';
      const result = formatIntlString(mockMessage);

      expect(intl.formatMessage).toHaveBeenCalledWith({ id: mockMessage }, undefined);
      expect(result).toBe(mockMessage);
    });

    it('should pass the message params as well', () => {
      const { intl, formatIntlString } = setup();

      const mockMessage = 'mock';
      const mockParams = {
        foo: 'bar',
      };

      formatIntlString(mockMessage, mockParams);
      expect(intl.formatMessage).toHaveBeenCalledWith({ id: mockMessage }, mockParams);
    });

    describe('error cases', () => {
      const cases: [string, null | undefined | string][] = [
        [ 'null', null ],
        [ 'undefined', undefined ],
        [ 'empty', '' ],
      ];

      describe.each(cases)('called with "%s"', (name, value) => {
        const { intl, formatIntlString } = setup();

        let consoleErrorSpy: jest.SpyInstance;

        beforeAll(() => {
          consoleErrorSpy = jest.spyOn(console, 'error');
        });

        afterAll(() => {
          consoleErrorSpy.mockClear();
        });

        it('should return the passed value as-is', () => {
          const result = formatIntlString(value as string);
          expect(result).toBe(value);
        });

        it('should not call intl.formatMessage', () => {
          expect(intl.formatMessage).not.toHaveBeenCalled();
        });

        it('should write an error on console.error', () => {
          expect(consoleErrorSpy).toHaveBeenCalledWith('[IntlForm] Invalid id given to formatIntlString');
        });
      });
    });
  });
});
