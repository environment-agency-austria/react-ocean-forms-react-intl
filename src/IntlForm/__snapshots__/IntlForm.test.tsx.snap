// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IntlForm /> should render without crashing 1`] = `
<Form
  asyncValidateOnChange={false}
  asyncValidationWait={400}
  defaultValues={Object {}}
  disabled={false}
  formatString={[Function]}
  intl={
    Object {
      "defaultFormats": null,
      "defaultLocale": "en",
      "formatDate": [MockFunction],
      "formatHTMLMessage": [MockFunction],
      "formatMessage": [MockFunction],
      "formatNumber": [MockFunction],
      "formatPlural": [MockFunction],
      "formatRelative": [MockFunction],
      "formatTime": [MockFunction],
      "formats": null,
      "locale": "en",
      "messages": Object {},
      "now": [MockFunction],
      "onError": [MockFunction],
    }
  }
  plaintext={false}
/>
`;
